# BEE 4750 Lab 4: Simulation-Optimization

**Name**: Elliot Walsh

**ID**: 5101573, ejw224

> **Due Date**
>
> Friday, 11/17/23, 9:00pm

## Setup

The following code should go at the top of most Julia scripts; it will
load the local package environment and install any needed packages. You
will see this often and shouldn’t need to touch it.

In [1]:
import Pkg
Pkg.activate(".")
Pkg.instantiate()

  Activating project at `~/Desktop/fall 2023/cee 4750/lab4/lab04-ejw224`


   Installed Distances ─────────── v0.10.10
   Installed PDMats ────────────── v0.11.28
   Installed Setfield ──────────── v1.1.1
   Installed Scratch ───────────── v1.2.1


   Installed ConcurrentUtilities ─ v2.3.0
   Installed TranscodingStreams ── v0.10.2
   Installed Roots ─────────────── v2.0.22


   Installed OpenSSL_jll ───────── v3.0.12+0
   Installed Combinatorics ─────── v1.0.2
   Installed CommonSolve ───────── v0.2.4


   Installed Qt6Base_jll ───────── v6.5.3+1
   Installed JMcDM ─────────────── v0.7.8
   Installed SearchSpaces ──────── v0.2.0
   Installed LaTeXStrings ──────── v1.3.1


   Installed FillArrays ────────── v1.7.0
   Installed Metaheuristics ────── v3.3.3
   Installed Distributions ─────── v0.25.103


   Installed FFMPEG_jll ────────── v4.4.4+1


Precompiling project...


  ✓ Scratch
  ✓ CommonSolve
  ✓ LaTeXStrings


  ✓ ConcurrentUtilities


  ✓ TranscodingStreams
  ✓ Distances
  ✓ Combinatorics


  ✓ FillArrays
  ✓ OpenSSL_jll


  ✓ PDMats


  ✓ JMcDM


  ✓ Setfield
  ✓ RelocatableFolders


  ✓ TranscodingStreams → TestExt
  ✓ QuadGK
  ✓ Distances → DistancesSparseArraysExt


  ✓ FillArrays → FillArraysSparseArraysExt
  ✓ SearchSpaces


  ✓ CodecZlib


  ✓ Distances → DistancesChainRulesCoreExt


  ✓ FillArrays → FillArraysStatisticsExt


  ✓ Qt6Base_jll


  ✓ Latexify


  ✓ StatsFuns → StatsFunsChainRulesCoreExt
  ✓ FFMPEG_jll
  ✓ OpenSSL


  ✓ Roots


  ✓ FFMPEG


  ✓ UnitfulLatexify


  ✓ GR_jll


  ✓ HTTP


  ✓ Distributions


  ✓ Distributions → DistributionsTestExt


  ✓ Distributions → DistributionsChainRulesCoreExt


  ✓ GR


  ✓ Metaheuristics


  ✓ Plots


  ✓ Plots → UnitfulExt
  38 dependencies successfully precompiled in 50 seconds. 147 already precompiled.


In [3]:
using Random # for random seeds
using Distributions # statistical distribution interface
using Roots # find zeros of functions
using Metaheuristics # search algorithms
using Plots # plotting

## Overview

In this lab, you will experiment with simulation-optimization with the
shallow lake problem. The goal of this experimentation is to get an
understanding of how to work with simulation-optimization methods and
the impact of some choices involved in using these methods.

Free free to delete some of the illustrative cells and code blocks in
your notebook as you go through and solve the lab problems…this might
help reduce some potential confusion while grading about what your
answer is.

## Introduction

Due to ongoing economic activity, a town emits phosphorous into a
shallow lake (with a concentration of $a_t$), which also receives
non-point source runoff (concentration $y_t$) from the surrounding area.
The concentration of the lake at time $t+1$ is given by
$$X_{t+1} = X_t + a_t + y_t + \frac{X_t^q}{1+X_t^q} - bX_t,$$

where:

| Parameter | Value                                                |
|:---------:|:-----------------------------------------------------|
|   $a_t$   | point-source phosphorous concentration from the town |
|   $y_t$   | non-point-source phosphorous concentration           |
|    $q$    | rate at which phosphorous is recycled from sediment  |
|    $b$    | rate at which phosphorous leaves the lake            |

and $X_0 = 0$, $y_t \sim LogNormal(\log(0.03), 0.25)$, $q=2.5$, and
$b=0.4$.

The goal of the optimization is to maximize the town’s average
phosphorous concentration releases (as a proxy of economic activity):
$\max \sum_{t=1}^T a_t / T$ over a 100-year period. We have decided
(initially) that an acceptable solution is one which will result in the
lake eutrophying no more than 10% of the time.

The non-point source samples can be sampled using the following code
block:

In [63]:
Random.seed!(1)

T = 100 # length of simualtion
n_samples = 1_000 # replace with number of samples if you experiment

P_distribution = LogNormal(log(0.03), 0.25)
y = rand(P_distribution, (T, n_samples)) # sample a T x n_samples matrix

100×1000 Matrix{Float64}:
 0.0294753  0.0459864  0.023513   …  0.0259183  0.0260934  0.0284652
 0.034263   0.0222782  0.0459188     0.0288482  0.0480454  0.0531018
 0.0245199  0.0296271  0.0445619     0.0246404  0.0250734  0.0304308
 0.055448   0.0312     0.0228208     0.0298609  0.0428105  0.0256198
 0.0401417  0.024978   0.0458244     0.0228935  0.0286062  0.0238694
 0.0320754  0.021759   0.0471452  …  0.0472771  0.0187508  0.0306753
 0.0464641  0.0416385  0.0246833     0.0382252  0.0288505  0.0226561
 0.0244027  0.0432707  0.0341214     0.0238988  0.0427204  0.0316143
 0.0231156  0.0279197  0.0217747     0.0231772  0.0335662  0.0324465
 0.0276303  0.0305858  0.0440326     0.0289394  0.0312328  0.0173388
 ⋮                                ⋱                        
 0.025665   0.0341366  0.0274747     0.0283546  0.0458031  0.0277959
 0.0405629  0.0421121  0.0252557     0.0450377  0.0284411  0.0206434
 0.0228445  0.0223746  0.0210942     0.0442834  0.0337672  0.0287835
 0.0252604  0.046

We write the lake model as a function:

In [30]:
# lake function model
# inputs:
#   a: vector of point-source releases (to be optimized)
#   y: randomly-sampled non-point sources
#   q: lake phosphorous recycling rate
#   b: phosphorous outflow rate
# 
# returns:
#   series of lake phosphorous concentrations
function lake(a, y, q, b, T)
    X = zeros(T+1, size(y, 2))
    # calculate states
    for t = 1:T
        X[t+1, :] = X[t, :] .+ a[t] .+ y[t, :] .+ (X[t, :].^q./(1 .+ X[t, :].^q)) .- b.*X[t, :]
    end
    return X
end

lake (generic function with 1 method)

However, this isn’t sufficient on its own! `Metaheuristics.jl` (and most
simulation-optimization packages) require the use of a *wrapper*
function, which accepts as inputs both parameters to be optimized (in
this case, point-source releases `a`) and parameters which will be fixed
(the others; see below for how to incorporate these into the syntax) and
returns the required information for the optimization procedure.

`Metaheuristics.jl` wants its optimizing wrapper function to return (in
order):

-   the objective(s) (in this case, the mean of `a`, $\sum_t a_t / T$),
-   a vector of the degrees to which the solution fails to achieve any
    inequality constraints (positive values indicate a larger failure,
    values below zero are considered acceptable)
-   a vector of the degrees to which the solution fails to achieve any
    equality constraints (only values of zero indicate success), which
    in this case is not relevant, so we just return `[0.0]`.

In [57]:
# function producing optimization outputs
# inputs:
#   a: vector of point-source releases (to be optimized)
#   y: randomly-sampled non-point sources
#   q: lake phosphorous recycling rate
#   b: phosphorous outflow rate
# 
# returns:
#   - objective: mean value of point-source releases
#   - inequality constraint failure vector
#   - equality constraint failure vector (in this case, always [0.0])
function lake_opt(a, y, q, b, T, Xcrit)
    X = lake(a, y, q, b, T)
    # calculate exceedance of critical value
    Pexceed = sum(X[T+1, :] .> Xcrit) / size(X, 2)
    failconst = [Pexceed - 0.1] # replace 0.1 if you experiment with the failure probability
    return mean(a), failconst, [0.0]
end

lake_opt (generic function with 1 method)

To optimize using DE (differential evolution), use the following syntax:

``` julia
results = optimize(f, bounds, DE(options=Options(f_calls_limit=max_evals)))
```

where `bounds` is a `Matrix` of lower bounds (first row) and upper
bounds (last row), and `max_evals` is an integer for the maximum number
of evaluations.

-   For example, to set bounds for all decision variables between 0 and
    0.5, you can use

``` julia
bounds = [zeros(T) 0.5ones(T)]
```

-   Increasing `max_evals` can help you find a better solution, but at a
    larger computational expense.
-   You can use an anonymous function to fix values for non-optimized
    parameters, *e.g.*

``` julia
y = ...
q = ...
b = ...
T = ...
Xcrit = ...
results = optimize(a -> lake_opt(a, y, q, b, t, Xcrit), bounds, DE(options=Options(f_calls_limit=max_evals)))
```

Then to get the approximated minimum value:

``` julia
fx = minimum(result)
```

and the approximated minimizing value:

``` julia
x = minimizer(result)
```

The last piece is to get the critical value (to identify failures),
which we can do using `Roots.jl`, which finds zeros of functions:

In [52]:
# define a function whose zeros are the critical values
P_flux(x) = (x^q/(1+x^q)) - b*x
# use Roots.find_zero() to find the non-eutrophication and non-zero critical value; we know from visual inspection in class that this is bounded between 0.1 and 1.5.
Xcrit = find_zero(P_flux, (0.1, 1.5))

0.6678778690448219

## Problems

### Problem 1 (2 points)

Using the default setup above, find the approximate optimum value. What
is the value of the objective function, and how many failures (you can
evaluate the `lake` function using your solution to find how many
end-values are above the critical value).

In [68]:
q=2.5; #phosphorus recycling rate
b=0.4 #phosphorus leaving rate

# define a function whose zeros are the critical values
P_flux(x) = (x^q/(1+x^q)) - b*x;
# use Roots.find_zero() to find the non-eutrophication and non-zero critical value; we know from visual inspection in class that this is bounded between 0.1 and 1.5.
Xcrit = find_zero(P_flux, (0.1, 1.5));

#a_bounds = [0.1ones(T) 1.5ones(T)]
a_bounds = [zeros(T) 0.1ones(T)]
max_evals=1000

results = optimize(a -> lake_opt(a, y, q, b, T, Xcrit), a_bounds, DE(options=Options(f_calls_limit=max_evals)))

Optimization Result
===================
  Iteration:       1
  Minimum:         0.0430547
  Minimizer:       [0.0619344, 0.0967516, 0.0561818, …, 0.0177579]
  Function calls:  1000
  Feasibles:       0 / 1000 in final population
  Total time:      3.5093 s
  Stop reason:     Maximum objective function calls exceeded.

The optimal value of the objective function (the cumulative point source total phosphorus input) is 0.208527. With bounds for a [0, 0.5], out of 1000 evaluations, none resulted in a \<10% chance of eutrophication (so 1000 were failures).

### Problem 2 (5 points)
Feel free to experiment with some of the degrees of freedom in finding
the optimum value. For example:

-   What failure probability are you using to characterize acceptable
    solutions?
-   How many Monte Carlo samples are you using?
-   What bounds are you searching over for the releases?
-   How many function evaluations are you using for the search?
-   What is the impact of different [`Metaheuristics.jl`
    algorithms](https://docs.juliahub.com/Metaheuristics/aJ70z/3.2.12/algorithms/)?

Note that you might want to modify some of these together: for example,
lower acceptable failure probabilities often require more function
evaluations to find acceptable values, and more Monte Carlo samples
increase computational expense, so fewer function evaluations may be
completed in the same time.

Provide a description of what you’ve modified and why. What was the new
solution that you found? Does it satisfy the constraints?

Even with 20,000 evaluations, none succeeded with the original parameters. Shortening the simulation length T to 10 increases the number of successes to 6/100, with an objective value of 0.0586. This makes sense - given less time to eutrophy, the lake is more likely to remain healthy.

Changing the bounds of a to [0, 0.1] (with 1000 evaluations) results in 6/1000 successes, with an objective value of 0.0395. Further reducing a's upper bound to 0.05 resulted in 1000/1000 successes, with an objective value of 0.019.

Returning to the original bounds of [0, 0.5] but changing the acceptable threshold to 20% resulted in 1000/1000 successes, with an objective value of 0.019. 


### Problem 3 (3 points)

What did you learn about the use of these methods? Compare with your
experience with linear programming from earlier in the semester.

I had never used this sort of optimization model before, so this lab was a novel experience. I found JuMP's way of defining parameters, constraints, and objectives more intuitive, but this approach is obviously much more powerful in many cases because of its ability to model nonlinear systems. Having Monte Carlo variation as well as multiple evaluation trials also posed a challenge, as it was easy to lose track of the original problem. Still, I have a better grasp now of how to implement this sort of model for an arbitrary optimization problem.

## References

Put any consulted sources here, including classmates you worked with/who
helped you.